onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,719 @@
+// frontier nodes require type change from FLOAT to FLOAT16. 
+// require_type_change consists of such nodes.  All the frontier nodes require fp32 cast
+// The input node is expected to be non-nullptr
+static bool PropagateForwards(Graph& graph, Node* node,
+                              std::deque<onnxruntime::NodeIndex>& removed_nodes,
+                              size_t level,
+                              const logging::Logger& logger)
+{
+  ORT_ENFORCE(node != nullptr);
+  bool modified = false;
+  if (IsCastTo(*node, TensorProto::FLOAT)) {
+    std::unordered_set<NodeArg*> require_cast;
+    std::unordered_set<NodeArg*> require_type_change;
+    NodeArg* cast_output = node->MutableOutputDefs()[0];
+    SearchDownstream(graph, cast_output, require_cast, require_type_change, removed_nodes, level);
+    if (require_cast.size() > 0 && require_cast.find(cast_output) == require_cast.end()) {
+      // Remove Cast operation
+      VLOGS(logger, 1) << "PropagateForwards: Removed Cast node  " << node->Name();
+      RemoveCastNodes(graph, {node}, removed_nodes);
+      InsertCastNodes(graph, require_cast, false, removed_nodes);
+      ChangeTypeToFP16(require_type_change, logger);
+      VLOGS(logger, 1) << "PropagateForwwards: Inserted Cast nodes " << ConcatNames<std::unordered_set<NodeArg*>>(require_cast);
+      modified = true;
+    }
+  } else if (node->OpType() != "Cast") {
+    for (NodeArg* output: node->MutableOutputDefs()) {
+      for (Node* consumer : graph.GetMutableConsumerNodes(output->Name())) {
+        if (consumer && std::find(removed_nodes.begin(), removed_nodes.end(), consumer->Index()) == removed_nodes.end()) {
+          modified |= PropagateForwards(graph, consumer, removed_nodes, level, logger);
+        }
+      }
+    }
+  }
+  return modified;
+}
+
+// PropagateBackwards
+// Propagate FP16 Cast operations backwards (upstream)
+// Recurrsively search the graph for Cast FP16 safe/allowed operations and expand
+// float16 computation regsion and
+// find the frontiers of the float16 computation region.
+// The required_cast vector is the collection of 
+// FP16-cast-frontiers of the cast node. All nodeargs on the path from any of the 
+// frontier nodes to the cast node require type change from  FLOAT to FLOAT16.
+// Each of the frontier nodes requires an fp16 cast.
+// The input node is expected be non-nullptr.
+static bool PropagateBackwards(Graph& graph, Node* node,
+                               std::deque<onnxruntime::NodeIndex>& removed_nodes,
+                               size_t level,
+                               const logging::Logger& logger)
+{
+  bool modified = false;
+  ORT_ENFORCE(node != nullptr);
+  if (IsCastTo(*node, TensorProto::FLOAT16)) {
+    std::unordered_set<NodeArg*> require_cast;
+    NodeArg* cast_input = node->MutableInputDefs()[0];
+    std::unordered_set<NodeArg*> require_type_change;
+    SearchUpstream(graph, cast_input, require_cast, require_type_change, removed_nodes, level);
```

Comment:
SearchUpstream itself is another recursive function. 
We are calling one recursive function insider another recursive function.... This is a bit nerve wracking...